### DISCUSSION TITLE
Always the same output of kekulization?

### DISCUSSION THREAD

**Discussion by najwalb** [rdkit/rdkit#6358] [Q&A] [answered]
I am using RDKit 2023-03-01.

I am generating kekule forms of the same molecules starting from different SMILES (i.e. resonance structures). I always end up with the same kekule form (in SMILES and images) no matter the input. 

- How does RDKit choose which Kekule form to use for each molecule?
- Is it guaranteed that RDKit will always output the same Kekule form given different resonance structures of the same molecule?

Here is an example:
```
    smiles = ['c1c(Cl)c(Cl)ccc1', 'C1C(Cl)=C(Cl)C=CC=1', 'C1=C(Cl)C(Cl)=CC=C1']
    for i, smi in enumerate(smiles):
        m = Chem.MolFromSmiles(smi)
        smi_kkl = Chem.MolToSmiles(m, kekuleSmiles=True)
        print(f'Original smi {smi}\n')
        print(f'Kekule smi   {smi_kkl}\n')

        img = Draw.MolToImage(m, kekulize=True)
        img.save(os.path.join(parent_path, 'figures', f'{smi}_{i}.png'))
```

The output is:
```
Original smi c1c(Cl)c(Cl)ccc1

Kekule smi   ClC1=CC=CC=C1Cl

Original smi C1C(Cl)=C(Cl)C=CC=1

Kekule smi   ClC1=CC=CC=C1Cl

Original smi C1=C(Cl)C(Cl)=CC=C1

Kekule smi   ClC1=CC=CC=C1Cl
```

I also asked [here](https://mattermodeling.stackexchange.com/questions/10884/cannot-draw-resonance-structure-kekule-forms-of-a-molecule-in-rdkit) and [here](https://github.com/rdkit/rdkit/issues/6354) about not being able to plot different resonance structures of the same molecule. I think the two questions are related.

**Comment by ptosco** [comment, marked as answer]
@najwalb By default, RDKit sanitizes molecules, and sanitization includes a kekulization and an aromatisation step. Therefore, the original position of single/double bonds which are part of an aromatic system is lost, and bond types are set to `AROMATIC`.
Upon visualization, by default the molecule is re-kekulized from its aromatic form as it looks nicer with localized double bonds, and alternate single/double bond positions are set by the RDKit kekulizer.
If you wish to preserve the original position of the single/double bonds when the molecule is generated, you need to switch off the automatic sanitization, and manually sanitize skipping the kekulization and aromatisation steps. To preserve the original location of single/double bonds you also need to switch off the automatic kekulization at visualization time.
The example below shows how do accomplish this:
```
import os
from rdkit import Chem
from rdkit.Chem.Draw import rdMolDraw2D
from IPython.display import Image

smiles = ['c1c(Cl)c(Cl)ccc1', 'C1C(Cl)=C(Cl)C=CC=1', 'C1=C(Cl)C(Cl)=CC=C1']
smiles_parser_params = Chem.SmilesParserParams()
smiles_parser_params.sanitize = False

def generate_images(smiles, should_kekulize_aromatic_smiles=False):
    imgs = []
    for i, smi in enumerate(smiles):
        m = Chem.MolFromSmiles(smi, smiles_parser_params)
        has_aromatic_atoms = False
        if should_kekulize_aromatic_smiles:
            has_aromatic_atoms = any(a.GetIsAromatic() for a in m.GetAtoms())
        sanitize_flags = Chem.SanitizeFlags.SANITIZE_ALL
        should_kekulize = should_kekulize_aromatic_smiles
        if not has_aromatic_atoms:
            sanitize_flags ^= (Chem.SanitizeFlags.SANITIZE_KEKULIZE | Chem.SanitizeFlags.SANITIZE_SETAROMATICITY)
            should_kekulize = False
        Chem.SanitizeMol(m, sanitize_flags)
        smi_kkl = Chem.MolToSmiles(m)
        print(f'Original smi {smi}\n')
        print(f'Kekule smi   {smi_kkl}\n')
        drawer = rdMolDraw2D.MolDraw2DCairo(200, 100)
        drawer.drawOptions().prepareMolsBeforeDrawing = False
        m = rdMolDraw2D.PrepareMolForDrawing(m, kekulize=should_kekulize)
        drawer.DrawMolecule(m)
        drawer.FinishDrawing()
        img = drawer.GetDrawingText()
        imgs.append(img)
        with open(os.path.join(parent_path, 'figures', f'{smi}_{i}.png'), "wb") as hnd:
            hnd.write(img)
    return imgs

imgs = generate_images(smiles)
Original smi c1c(Cl)c(Cl)ccc1

Kekule smi   Clc1ccccc1Cl

Original smi C1C(Cl)=C(Cl)C=CC=1

Kekule smi   ClC1=C(Cl)C=CC=C1

Original smi C1=C(Cl)C(Cl)=CC=C1

Kekule smi   ClC1=CC=CC=C1Cl

Image(imgs[0])
```
![c1c(Cl)c(Cl)ccc1_0](https://github.com/rdkit/rdkit/assets/5244385/c6aa52ee-b04e-4df7-ae43-2c40827edd5b)
```
Image(imgs[1])
```
![C1C(Cl)=C(Cl)C=CC=1_1](https://github.com/rdkit/rdkit/assets/5244385/66f6816e-870c-4869-8938-64b42b6eefa1)
```
Image(imgs[2])
```
![C1=C(Cl)C(Cl)=CC=C1_2](https://github.com/rdkit/rdkit/assets/5244385/77d59dbf-be93-46c2-b476-37c2e72d1202)

If you prefer a kekulized visualization of aromatic SMILES, set `should_kekulize_aromatic_smiles` to `True`.
Note that this will kekulize the whole molecule, including other rings with pre-localized single/double bonds, whose original location will therefore be reset:
```
imgs = generate_images(smiles, True)
Original smi c1c(Cl)c(Cl)ccc1

Kekule smi   Clc1ccccc1Cl

Original smi C1C(Cl)=C(Cl)C=CC=1

Kekule smi   ClC1=C(Cl)C=CC=C1

Original smi C1=C(Cl)C(Cl)=CC=C1

Kekule smi   ClC1=CC=CC=C1Cl

Image(imgs[0])
```
![c1c(Cl)c(Cl)ccc1_0](https://github.com/rdkit/rdkit/assets/5244385/d8ea88cb-10d8-4113-a74d-ddacffc2cc41)
```
Image(imgs[1])
```
![C1C(Cl)=C(Cl)C=CC=1_1](https://github.com/rdkit/rdkit/assets/5244385/21aa3d22-426c-40b8-ace9-00d4e3a8f705)
```
Image(imgs[2])
```
![C1=C(Cl)C(Cl)=CC=C1_2](https://github.com/rdkit/rdkit/assets/5244385/1e572d66-0072-4acc-8cf9-bb7d697c82e1)

**Reply by najwalb** [reply]
Perfect thanks a lot! @ptosco I still need a little help with getting the right SMILES from ResonanceMolSupplier using this idea of turning off sanitization. I am able to retrieve the right images with a small modifications of your `generate_images()` function, but no success with SMILES so far.

Here is the code:

```
def generate_images(mols, should_kekulize_aromatic_smiles=False):
    imgs = []
    smiles_parser_params = Chem.SmilesParserParams()
    smiles_parser_params.sanitize = False
    
    for i, m in enumerate(mols):
        smi = Chem.MolToSmiles(m)
        has_aromatic_atoms = False
        if should_kekulize_aromatic_smiles:
            has_aromatic_atoms = any(a.GetIsAromatic() for a in m.GetAtoms())
        sanitize_flags = Chem.SanitizeFlags.SANITIZE_ALL
        should_kekulize = should_kekulize_aromatic_smiles
        if not has_aromatic_atoms:
            sanitize_flags ^= (Chem.SanitizeFlags.SANITIZE_KEKULIZE | Chem.SanitizeFlags.SANITIZE_SETAROMATICITY)
            should_kekulize = False

        Chem.SanitizeMol(m, sanitize_flags)
        smi = Chem.MolToSmiles(m, kekuleSmiles=False)
        smi_kkl = Chem.MolToSmiles(m, kekuleSmiles=should_kekulize)
        print(f'Original smi {smi}\n')
        print(f'Kekule smi   {smi_kkl}\n')
        drawer = rdMolDraw2D.MolDraw2DCairo(200, 100)
        drawer.drawOptions().prepareMolsBeforeDrawing = False
        m = rdMolDraw2D.PrepareMolForDrawing(m, kekulize=should_kekulize)
        drawer.DrawMolecule(m)
        drawer.FinishDrawing()
        img = drawer.GetDrawingText()
        imgs.append(img)
        with open(os.path.join(parent_path, 'figures', f'{smi_kkl}_{i}.png'), "wb") as hnd:
            hnd.write(img)
    return imgs
```

```
smi = 'c1c(Cl)c(Cl)ccc1'
mol = Chem.MolFromSmiles(smi)
suppl = Chem.ResonanceMolSupplier(mol, Chem.KEKULE_ALL)
mols = [mol]+[m for m in suppl]
imgs = generate_images(mols=mols, should_kekulize_aromatic_smiles=False)
```
And the output:
```

Original smi Clc1ccccc1Cl

Kekule smi   Clc1ccccc1Cl

Original smi Clc1ccccc1Cl

Kekule smi   Clc1ccccc1Cl

Original smi Clc1c(Cl)cccc1

Kekule smi   Clc1c(Cl)cccc1
```

Technically in this case I want to only print original smi but in the same form you had for kekule smi before (i.e.  `Clc1ccccc1Cl`, `C1C(Cl)=C(Cl)C=CC=1` and `ClC1=CC=CC=C1Cl`), 

The output for images works fine:

`Image(imgs[0])`

![Clc1ccccc1Cl_2](https://github.com/rdkit/rdkit/assets/107994360/da082a69-d14d-4fe5-a31f-136183a5a5a5)

`Image(imgs[1])`

![Clc1c(Cl)cccc1_1](https://github.com/rdkit/rdkit/assets/107994360/e25d0a5c-9cc1-4cc3-b269-71bb353ff68b)

`Image(imgs[2])`

![Clc1ccccc1Cl_0](https://github.com/rdkit/rdkit/assets/107994360/a87a55a7-8911-4180-8c1e-a9a1b7a2b323)

**Comment by ptosco** [comment]
@najwalb You need to unset the aromatic flags to obtain the Kekule SMILES:
```
def unset_aromatic_flags(m):
    for b in m.GetBonds():
        if b.GetBondType() != Chem.BondType.AROMATIC and b.GetIsAromatic():
            b.SetIsAromatic(False)
            b.GetBeginAtom().SetIsAromatic(False)
            b.GetEndAtom().SetIsAromatic(False)
    return m

smi = 'c1c(Cl)c(Cl)ccc1'
mol = Chem.MolFromSmiles(smi)
suppl = Chem.ResonanceMolSupplier(mol, Chem.ResonanceFlags.KEKULE_ALL)
mols = [mol]+[unset_aromatic_flags(m) for m in suppl]

imgs = generate_images(mols=mols, should_kekulize_aromatic_smiles=False)
Original smi Clc1ccccc1Cl

Kekule smi   Clc1ccccc1Cl

Original smi ClC1=CC=CC=C1Cl

Kekule smi   ClC1=CC=CC=C1Cl

Original smi ClC1=C(Cl)C=CC=C1

Kekule smi   ClC1=C(Cl)C=CC=C1
```

**Comment by najwalb** [comment]
Perfect it's working now, thanks a lot!

**Comment by LingjieBao1998** [comment]
Perfect it's working now, thanks a lot! it helps me a lot.
